### ISSUE TITLE
ARM 32-bit PUSH Emulation Error

### ISSUE THREAD

**Issue by SpenserStyles** [radare/radare2#10632]
### Work environment

| Questions                                            | Answers
|------------------------------------------------------|--------------------
| OS/arch/bits (mandatory)                             | Mac 10.13.5 64-bit
| File format of the file you reverse (mandatory)      | BIN
| Architecture/bits of the file (mandatory)            | arm 32-bit cortex
| r2 -v full output, **not truncated** (mandatory)     | radare2 2.6.9 18640 @ darwin-x86-64 git.2.6.9 commit: 159d16b34550f4a83ee496839981c8316339ac4d build: 2018-07-03__09:48:34
### Actual behavior
Push instruction push highest register to lowest register onto stack.  Lowest register is at lowest address (last register to be pushed)

PUSH {r1,lr} => LR is pushed first then R1

### Steps to reproduce the behavior 
Emulate push instruction for ARM 32-bit Cortex-M

I fixed this by simply changing the order of how the registers are written for esil.

```
case ARM_INS_PUSH:
#if 0
PUSH { r4, r5, r6, r7, lr }
4,sp,-=,lr,sp,=[4],
4,sp,-=,r7,sp,=[4],
4,sp,-=,r6,sp,=[4],
4,sp,-=,r5,sp,=[4],
4,sp,-=,r4,sp,=[4]

20,sp,-=,r4,r5,r6,r7,lr,5,sp,=[*]
#endif
		r_strbuf_setf (&op->esil, "%d,sp,-=,",
			4 * insn->detail->arm.op_count);

                // original  incorrect method for creating ESIL
		// for (i=0; i<insn->detail->arm.op_count; i++) {
		// 	r_strbuf_appendf (&op->esil, "%s,", REG (i));
		// }
	
		// Fix for ARM Cortex CPU to push things onto stack in correct order
		// push {r1,r2,lr}  => PUSH {LR} , PUSH{R2}, PUSH{R1}
		for (i=insn->detail->arm.op_count; i>0; i--) {
			r_strbuf_appendf (&op->esil, "%s,", REG (i-1));
		}

```

**Comment by radare** [comment]
If you fix something send a pullreq

**Comment by Maijin** [comment]
Use https://help.github.com/articles/editing-files-in-your-repository/

**Comment by astuder** [comment]
I think this is currently broken on a more fundamental level of ESIL. Only the last register is stored on the stack, while the rest of the stack is filled with mostly 0's and some random garbage.

```
[0x40220338]> pd 1
            ;-- fcn.40220338:
            ;-- pc:
            ;-- r15:
╭ (fcn) fcn.c_init_nonzero 178
│   fcn.c_init_nonzero (int32_t arg1, int32_t arg_0h, int32_t arg_4h, int32_t arg_8h);
│ bp: 0 (vars 0, args 0)
│ sp: 3 (vars 0, args 3)
│ rg: 1 (vars 0, args 1)
│           0x40220338  ~   feb5           push {r1, r2, r3, r4, r5, r6, r7, lr}
[0x40220338]> px 0x60 @ sp-0x30
- offset -   0 1  2 3  4 5  6 7  8 9  A B  C D  E F  0123456789ABCDEF
0x40b77fc8  0000 0000 0000 0000 0000 0000 0000 0000  ................
0x40b77fd8  0000 0000 0000 0000 0000 0000 0000 0000  ................
0x40b77fe8  0000 0000 0000 0000 0000 0000 68f9 2240  ............h."@
0x40b77ff8  6e00 0000 0000 0000 0000 0000 0000 0000  n...............
0x40b78008  0000 0000 7700 0000 7c00 0000 0000 0000  ....w...|.......
0x40b78018  0000 0000 004b 219a 827f 0000 0500 0000  .....K!.........
[0x40220338]> aes
[0x40220338]> px 0x60 @ sp-0x10
- offset -   0 1  2 3  4 5  6 7  8 9  A B  C D  E F  0123456789ABCDEF
0x40b77fc8  0000 0000 0000 0000 0000 0000 0000 0000  ................
0x40b77fd8  a029 2240 6e00 0000 0000 0000 0000 0000  .)"@n...........
0x40b77fe8  0000 0000 0000 0000 7700 0000 7c00 0000  ........w...|...
0x40b77ff8  0000 0000 0000 0000 004b 219a 827f 0000  .........K!.....
0x40b78008  0500 0000 0000 0000 5079 b0c9 fe55 0000  ........Py...U..
0x40b78018  0000 0000 6e00 0000 0000 0000 0000 0000  ....n...........
[0x40220338]> ar
sb = 0x00000000
sl = 0x00000000
fp = 0x40b78000
ip = 0x00000000
sp = 0x40b77fd8
lr = 0x402229a0
pc = 0x4022033a
r0 = 0x4022f930
r1 = 0x40220339
r2 = 0xffff0a0d
r3 = 0x000191a0
r4 = 0x4022f968
r5 = 0x00000000
r6 = 0x00000000
r7 = 0x00000000
r8 = 0x00000000
r9 = 0x00000000
r10 = 0x00000000
r11 = 0x40b78000
r12 = 0x00000000
r13 = 0x40b77fd8
r14 = 0x402229a0
r15 = 0x4022033a
cpsr = 0x80000000
blank = 0x00000000
[0x40220338]> ao @ pc-2
address: 0x40220338
opcode: push {r1, r2, r3, r4, r5, r6, r7, lr}
disasm: push {r1, r2, r3, r4, r5, r6, r7, lr}
pseudo: push (r1, r2, r3, r4, r5, r6, r7, lr) 
mnemonic: push
mask: ffff
prefix: 0
id: 424
bytes: feb5
refptr: 0
size: 2
sign: false
type: push
cycles: 1
esil: 32,sp,-=,r1,r2,r3,r4,r5,r6,r7,lr,8,sp,=[*]
family: cpu
[0x40220338]> 
```
This is with the latest build as of today:
radare2 3.7.1 22559 @ linux-x86-64 git.1.5.0-7703-g6459cd38b
commit: 6459cd38bdc0b378c491396723188e41cdff0745 build: 2019-09-01__20:54:22

**Comment by astuder** [comment]
Fixed the stack corruption issue (pull #14948), still more work to be done for the original issue.
